feat(rebuild): adding rebuild I/O verification

dsavitskiy opened the pull request.

dsavitskiy commented:
bors try

dsavitskiy commented:
bors try

dsavitskiy commented:
bors try

tiagolobocastro commented on Jenkinsfile:
```diff
@@ -140,6 +140,7 @@ pipeline {
           anyOf {
             branch 'master'
             branch 'release/*'
+            branch 'trying'
```
Did you mean to make this change?

dsavitskiy replied:
Yes I did. Should I do a separate PR for it?

tiagolobocastro replied:
Ah ok.
Should we skip just the submodule check rather than the entire lint? Reason for asking is that having the lint in place helps make sure code is correctly linted which makes core review easier as they won't be able to test changes on CI if code is not linted properly.

dsavitskiy replied:
I think, that's would be a better solution.

tiagolobocastro commented on io-engine/src/core/block_device.rs:
```diff
@@ -170,30 +158,184 @@ pub trait BlockDeviceHandle {
+            }
+        })?;
+
+        self.read_buf_blocks_async(
```
Can't we use compare command? This way we don't need to receive any data and the target will compare it for us.

dsavitskiy replied:
Yes we can. I try to use it.

dsavitskiy replied:
Done. I had to made a fix to SPDK to make compare APIs work properly: https://github.com/openebs/spdk/pull/35.

tiagolobocastro replied:
Yes I saw, approved that one already.

tiagolobocastro commented on io-engine/src/rebuild/rebuild_task.rs:
```diff
@@ -110,59 +111,137 @@ impl RebuildTask {
-                Some(ReadMode::UnwrittenFail),
-            )
-            .await;
+        let mut buf = src
```
Why are we now allocating the buffer here? We have to allocate/deallocate every time, whilst before we ensured we had the buffers for the duration of the rebuild.

dsavitskiy replied:
Fixed and improved to prevent buffer allocation for the final segment.

dsavitskiy committed:
feat(core): adding generic fault injection support

Fault injections moved from nexus to core.
Support for block-device level injection added.
New 'testing' gRPC service added.
Injections removed from v0 API.
CLI is changed accordingly.

Signed-off-by: Dmitry Savitskiy <dmitry.savitskiy@datacore.com>

dsavitskiy committed:
fix(ci): allowing bors try when submodule PR is not merged yet

Signed-off-by: Dmitry Savitskiy <dmitry.savitskiy@datacore.com>

dsavitskiy committed:
feat(core): improving block device abstraction interfaces

Changes to `BlockDeviceHandle` trait:
* `readv_blocks` and `writev_blocks` now accept `IoVec` slice instead of
  C-array of iovs and its length.
* `ReadOptions` argument is now passed to `readv_blocks` instead of `read_at`.
* New async methods added:
  - `readv_blocks_async`
  - `read_buf_blocks_async`
  - `writev_blocks_async`
  - `write_buf_blocks_async`
  These methods have default implementions in `BlockDeviceHandle`, based on
  `readv_blocks` and `writev_blocks` calls.
* `read_at` and `write_at` are now deprecated.

Rebuild task is changed to use `BlockDeviceHandle` methods.

Signed-off-by: Dmitry Savitskiy <dmitry.savitskiy@datacore.com>

dsavitskiy committed:
feat(rebuild): adding rebuild I/O verification

At this moment rebuild verification is controlled via NEXUS_REBUILD_VERIFY
shell var:
- "fail": rebuild job fails if verification fails
- "panic": I/O engine panics if verification fails
- any other value or not set: do no verify

Signed-off-by: Dmitry Savitskiy <dmitry.savitskiy@datacore.com>

dsavitskiy commented:
bors try

dsavitskiy commented:
bors try

dsavitskiy commented:
bors try

tiagolobocastro commented on io-engine/src/rebuild/rebuild_task.rs:
```diff
@@ -110,58 +123,21 @@ impl RebuildTask {
-        let mut copy_buffer: DmaBuf;
-        let source_hdl = descriptor.src_io_handle().await?;
-        let destination_hdl = descriptor.dst_io_handle().await?;
+        let buf_lock = self.buffer.lock().await;
```
Why does the buffer need a lock? we have a mutable ref here so we can modify it with extra indirection?

dsavitskiy replied:
Ah, indeed we don't need a lock here. I didn't notice that each task is already locked when called from RebuildTasks. Will fix.

tiagolobocastro commented on io-engine/src/rebuild/rebuild_descriptor.rs:
```diff
@@ -85,9 +100,168 @@ impl RebuildDescriptor {
+    ) -> IoVec {
+        let mut iov = buffer.to_io_vec();
+
+        let iov_len = self.get_segment_size_blks(offset_blk) * self.block_size;
```
Do we need to do this if we now specify the num_blocks ? on read/writev_blocks_async?

dsavitskiy replied:
It turned out that yes we have. I tried not doing that at first, and I/O failed in a strange way. I didn't look further but it looks like SPDK's APIs require num_blocks to be equal to iovecs size.

tiagolobocastro replied:
Ah I see, maybe that's the reason it was this way to begin with :(

tiagolobocastro commented on io-engine/src/grpc/v1/test.rs:
```diff
@@ -0,0 +1,172 @@
+    }
+}
+
+#[cfg(feature = "fault-injection")]
```
hmm this is not going to work very well if we want other things in testing in addition to fault injection, example: replica wiping and perhaps others in the future?

dsavitskiy replied:
This block is for injection only. Any other kinds of APIs can be added outside this block, with their own feature config if they need to be enabled selectively.
There's another feature "io-engine-testing", which included "fault-injection". Another test sub-features can be added there.

tiagolobocastro replied:
But traits have to be implemented as a whole right?

dsavitskiy replied:
ah, yes indeed. I'll change to per-method cfg condition.

tiagolobocastro replied:
You could add indirection by implementing injection as separate trait or just fn impls, and then in the TestRpc trait impl call into this injection fn's, example:
```rust
#[cfg(feature = injection)]
impl Injection { fn a { } fn b() { } } 
#[cfg(feature( no( injection))]
impl Injection { fn a { Not_impl } fn b() { Not_impl } }

impl TestRcp for TestService {
  fn a() { Injection::a() }
  fn b() { Injection::b() }
}
```

dsavitskiy commented:
bors try

tiagolobocastro commented on io-engine/src/bdev/nexus/nexus_bdev_rebuild.rs:
```diff
@@ -151,6 +153,31 @@ impl<'n> Nexus<'n> {
         src_child_uri: &str,
         dst_child_uri: &str,
     ) -> Result<(), Error> {
+        let verify_mode = match std::env::var("NEXUS_REBUILD_VERIFY")
```
Could this be a CLI argument?
Ex: sudo io-engine --rebuild-verify-ios

dsavitskiy replied:
We would end up with too many CLI args. Also, we were planned to make this a parameter of rebuild start / child online API.

tiagolobocastro replied:
We can have a separate struct for Testing cli arguments, then it's clear any "user" would ignore it but it will make it visible for testing what's available.

dsharma-dc commented on io-engine/src/rebuild/rebuild_descriptor.rs:
```diff
@@ -45,20 +54,23 @@ impl RebuildDescriptor {
     }
 
     /// Get a `BlockDeviceHandle` for the source.
+    #[inline(always)]
```
Is it a good idea to annotate `async fn` as `inline always` in Rust? or better leave it to compiler's wisdom.

dsavitskiy replied:
I guess, it's the same as for C++ (Rust and C++ share the same LLVM backend): it may probably help to force the compiler to inline.

dsharma-dc replied:
For C/C++ too unless we know function is getting called thousands of times per second and is self contained like a macro, we'd usually leave it to compiler as compilers today will do inlining even if we don't ask. Benchmarking would be the real measure of such impacts on instruction ordering and branching. I'd think it's superfluous here for these async functions.

dsharma-dc commented on io-engine-tests/src/test.rs:
It'd be good to use a uniformly used `fault_injection` everywhere instead of simply `injection`. e.g `pub async fn add_fault_injection`

dsavitskiy replied:
Fixed

dsharma-dc commented on io-engine/src/rebuild/rebuild_descriptor.rs:
```diff
@@ -183,4 +185,83 @@ impl RebuildDescriptor {
+        offset_blk: u64,
+        iovs: &mut [IoVec],
+    ) -> Result<(), RebuildError> {
+        // Read the source again.
```
Why do we need to read source again? The LBA ranges are still locked I believe when `verify_segment` is called and the buffers that we used to write to destination can be still valid.

dsavitskiy replied:
There can be source read errors.

dsharma-dc replied:
hmm, not sure if that should matter here. IIUC, the intent is to read from destination, and compare to the buffer from which we wrote.

tiagolobocastro commented on io-engine/src/rebuild/rebuild_descriptor.rs:
```diff
@@ -85,9 +100,168 @@ impl RebuildDescriptor {
+                source: err,
+                bdev: self.dst_uri.clone(),
+            }),
+        }
```
hmm I think this might yield false miscompares for thin volumes.
Let's say we have a thin volume and we're rebuilding a new replica.
1. want to rebuild 64k bytes which targets an unallocated cluster.
2. 64k range is locked
3. special read tells us we can skip the write
4. we want to now verify 
5. a user IO which targets a different range in the cluster means the cluster now is allocated
6. read from source
7. verify target might fail because allocating the cluster in 5 means now we have undefined data here

tiagolobocastro replied:
All, please ignore this, we don't verify if read was skipped so we never end up hitting this scenario!

dsavitskiy replied:
We don't verify if we don't write.

dsavitskiy commented:
bors merge

dsavitskiy commented:
bors merge

dsavitskiy commented:
bors merge

dsavitskiy commented:
bors merge

dsavitskiy commented:
bors merge
